public/apps/configuration/panels/role-edit.tsx

```diff
@@ -47,24 +62,289 @@ const TITLE_TEXT_DICT = {
+}
+
+/**
+ * Remove the leading ~ which indicates exclude and convert to combo box option.
+ * @param fieldLevelSecurityRawFields fields fetched from backend
+ * ["~field1", "~field2"] => ["field1", "field2"]
+ * ["field1", "field2"] => ["field1", "field2"]
+ */
+function getFieldLevelSecurityFields(fieldLevelSecurityRawFields: string[]): OptionSeletion {
+  return fieldLevelSecurityRawFields
+    .map((s: string) => s.replace(/^~/, ''))
+    .map(stringToComboBoxOption);
+}
+
+function buildIndexPermissionState(
+  indexPerm: RoleIndexPermission[]
+): RoleIndexPermissionStateClass[] {
+  return indexPerm.map(perm => ({
+    indexPatterns: perm.index_patterns.map(stringToComboBoxOption),
+    docLevelSecurity: perm.dls,
+    fieldLevelSecurityMethod: getFieldLevelSecurityMethod(perm.fls),
+    fieldLevelSecurityFields: getFieldLevelSecurityFields(perm.fls),
+    maskedFields: [],
+    allowedActions: [],
+  }));
+}
+
+const FIELD_LEVEL_SECURITY_PLACEHOLDER = `{
+    "bool": {
+        "must": {
+            "match": {
+                "genres": "Comedy"
+            }
+        }
+    }
+}`;
 
-function buildPermissionOptions(optionsList: string[]) {
-  return optionsList.map(e => ({ label: e }));
+function IndexPatternRow(props: {
+  value: OptionSeletion;
+  onChangeHandler: (s: OptionSeletion) => void;
+  onCreateHandler: (s: string) => void;
+}) {
+  return (
+    <FormRow headerText="Index" helpText="Specify index pattern using *">
+      <EuiComboBox
+        noSuggestions
+        placeholder="Search for index name or type in index pattern"
+        selectedOptions={props.value}
+        onChange={props.onChangeHandler}
+        onCreateOption={props.onCreateHandler}
+      />
+    </FormRow>
+  );
+}
+
+function IndexPermissionRow(props: {
+  value: OptionSeletion;
+  permisionOptionsSet: OptionSeletion;
+  onChangeHandler: (s: OptionSeletion) => void;
+}) {
+  return (
+    <FormRow
+      headerText="Index permissions"
+      headerSubText="You can specify permissions using both action groups or single permissions. 
+        An permission group is a list of single permissions.
+        You can often achieve your desired security posture using some combination of the default permission groups. 
+        You can also create your own reusable permission groups."
+    >
+      <EuiFlexGroup>
+        <EuiFlexItem style={{ maxWidth: '400px' }}>
+          <EuiComboBox
+            placeholder="Search for action group name or permission name"
+            options={props.permisionOptionsSet}
+            selectedOptions={props.value}
+            onChange={props.onChangeHandler}
+          />
+        </EuiFlexItem>
+        <EuiFlexItem grow={false}>
+          <EuiButton>Browse and select</EuiButton>
+        </EuiFlexItem>
+        <EuiFlexItem grow={false}>
+          <EuiButton iconType="popout" iconSide="right">
+            Create new permission group
```

Comment:
Minor: Looks like from the UI mockup. It says "action" not "permission".

Reply:
It says permission, my first implementation is wrong.